framework/include/chaincontrols/ChainControl.h

```diff
@@ -0,0 +1,146 @@
+ * Control that additionally provides the capability to produce/consume data values,
+ * to allow control operations to be chained together.
+ */
+class ChainControl : public Control
+{
+public:
+  static InputParameters validParams();
+
+  ChainControl(const InputParameters & parameters);
+
+  /**
+   * Initialization that occurs in \c ChainControlSetupAction, right before the dependencies
+   * are added.
+   *
+   * Note that the \c initialSetup() method is executed after all
+   * actions are completed, so it is too late to add control data dependencies.
+   */
+  virtual void init() {}
+
+  /**
+   * Returns the ChainControls that must run before this one
+   */
+  const std::vector<std::string> & getChainControlDataDependencies() const
+  {
+    return _control_data_depends_on;
+  }
+
+protected:
+  /**
+   * Declares chain control data with the given name and type
+   *
+   * @tparam T          Type of chain control data
+   * @param data_name   Chain control data name
+   * @param apply_object_prefix   If true, apply the object name as a prefix to the data name
+   */
+  template <typename T>
+  T & declareChainControlData(const std::string & data_name, bool apply_object_prefix = true);
+
+  /**
+   * Get a reference to control data that are specified in the input parameter 'param_name'
+   *
+   * @tparam T      Type of chain control data
+   * @param param   Parameter containing chain control data name
+   */
+  template <typename T>
+  const T & getChainControlData(const std::string & param);
+
+  /**
+   * Get a reference to control data value from a previous time step that is specified in the input
+   * parameter 'param_name'
+   *
+   * @tparam T      Type of chain control data
+   * @param param   Parameter containing chain control data name
+   */
+  template <typename T>
+  const T & getChainControlDataOld(const std::string & param);
+
+  /**
+   * Get a reference to control data that are specified by 'data_name' name
+   *
+   * @tparam T          Type of chain control data
+   * @param data_name   Chain control data name
+   */
+  template <typename T>
+  const T & getChainControlDataByName(const std::string & data_name);
+
+  /**
+   * Get a reference to control data value from previous time step that is specified by 'data_name'
+   * name
+   *
+   * @tparam T          Type of chain control data
+   * @param data_name   Chain control data name
+   */
+  template <typename T>
+  const T & getChainControlDataOldByName(const std::string & data_name);
+
+  /**
+   * Adds a chain control data dependency into the list
+   *
+   * @param data_name   Name of chain control data that this control depends upon
+   */
+  void addChainControlDataDependency(const std::string & data_name);
+
+  /// List of chain control data that this control depends upon
+  std::vector<std::string> _control_data_depends_on;
+};
+
+template <typename T>
+T &
+ChainControl::declareChainControlData(const std::string & data_name, bool apply_object_prefix)
```

Comment:
THM has this action `PostprocessorAsControlAction` that automatically adds a control data for each PP that gets added (but only those in the input file 🤮 ), and in THM the control `CopyPostprocessorValueControl` always named  the data to be the same as the PP. In the chain control `GetPostprocessorChainControl` I give the option to name it one way or the other. The idea was backwards compatibility with THM.

However, I don't think I personally like `PostprocessorAsControlAction`, since I'd rather be explicit about introducing control data into the system, even though it can be less convenient, so I'm open to trying to get rid of it in THM.

That said, I don't know if we should force the naming convention on applications, even though we think it wise (and having the prefix is the default at least).

Reply:
>  backwards compatibility with THM.

that is a good idea. If the chainControl does not get adopted by THM and the apps, we've created some debt by duplicating the system

> However, I don't think I personally like PostprocessorAsControlAction

me neither. Adding functor support to the ParsedControl could help you support PPs without a ChainControl intermediate object ?

Reply:
Yeah `ParsedControl` should be able to be extended to include PPs. I don't know yet if that would be using functors or not. I get confused for contexts that are not on an element/side/etc.

For now, `GetPostprocessorChainControl` gets the job done.